src/common/utxobased/keymanager/cleaners.ts

```diff
@@ -0,0 +1,164 @@
+import {
+  asCodec,
+  asMaybe,
+  asNumber,
+  asObject,
+  asOptional,
+  asString,
+  Cleaner
+} from 'cleaners'
+
+import {
+  asCurrencyFormat,
+  CurrencyFormat,
+  EngineInfo,
+  NetworkEnum
+} from '../../plugin/types'
+import { deriveXpubsFromKeys } from '../engine/utils'
+
+// We default to bip32 always for legacy reasons
+const asOptionalCurrencyFormat = asOptional(asCurrencyFormat, 'bip32')
+
+/**
+ * A cleaner for the private key format following the key-formats specification.
+ *
+ * (spec: https://github.com/EdgeApp/edge-core-js/blob/master/docs/key-formats.md)
+ */
+export interface PrivateKey {
+  coinType: number
+  format: CurrencyFormat
+  seed: string
+}
+export function asPrivateKey(
+  coinName: string,
+  coinType: number = 0
+): Cleaner<PrivateKey> {
+  return asCodec(
+    raw => {
+      if (raw == null || typeof raw !== 'object') {
+        throw new TypeError('Private keys must be objects')
+      }
+      const asCoinType = asOptional(asNumber, coinType)
+      return {
+        coinType: asCoinType(raw.coinType),
+        format: asOptionalCurrencyFormat(raw.format),
+        seed: asString(raw[`${coinName}Key`])
+      }
+    },
+    clean => {
+      const { coinType, format, seed } = clean
+      return {
+        coinType,
+        format,
+        [`${coinName}Key`]: seed
+      }
+    }
+  )
+}
+
+/**
+ * A cleaner for the public key format following the key-formats specification.
+ *
+ * (spec: https://github.com/EdgeApp/edge-core-js/blob/master/docs/key-formats.md)
+ */
+export interface PublicKey {
+  publicKeys: {
+    [format in CurrencyFormat]?: string
+  }
+}
+export const asPublicKey: Cleaner<PublicKey> = asObject({
+  publicKeys: asObject({
+    bip32: asOptional(asString),
+    bip44: asOptional(asString),
+    bip49: asOptional(asString),
+    bip84: asOptional(asString)
+  })
+})
+
+/**
+ * This cleaner returns the wallet's supported formats as specified in the
+ * key-formats specification.
+ */
+export const asWalletKeysFormats: Cleaner<CurrencyFormat[]> = (
+  keys: any
+): CurrencyFormat[] => {
+  const format = asOptionalCurrencyFormat(keys?.format)
+
+  switch (format) {
+    case 'bip32':
+      return ['bip32']
+    case 'bip44':
+      return ['bip44']
+    case 'bip49':
+      return ['bip49', 'bip84']
+    case 'bip84':
+      return ['bip49', 'bip84']
+    default:
+      throw new Error(`Unsupported format ${format}`)
+  }
+}
```

Comment:
This should not be a cleaner - it is being used in contexts where data has already been cleaned, so it should just be a normal utility function taking `CurrencyFormat` and returning `CurrencyFormat[]`. Let's call it `getSupportedFormats`, for the sake of this PR.